text/0000-secret-types.md

```diff
@@ -0,0 +1,206 @@
+- Overflowing_mul
+- Overflowing_neg
+- Overflowing_shl
+- Overflowiing_shr
+- Overflowing_pow
+- Pow
+- Reverse_bits
+- Rotate_left
+- Rotate_right
+- Saturating_add
+- Saturating_sub
+- Saturating_neg
+- Saturating_mul
+- Saturating_pow
+- Signum
+- Swap_bytes
+- To_be
+- To_le
+- To_be_bytes
+- To_le_bytes
+- To_ne_bytes
+- Trailing_zeros
+- Wrapping_add
+- Wrapping_sub
+- Wrapping_mul
+- Wrapping_neg
+- Wrapping_shl
+- Wrapping_shr
+- Wrapping_pow
+
+
+Implement the following traits
+Secret integers may only  be combined with other secret integers and the result will be a secret type
+- Add
+- AddAssign
+- BitAnd
+- BitAndAssign
+- BitOr
+- BitOrAssign
+- BitXor
+- BitXorAssign
+- Clone
+- Copy
+- Default
+- Drop
+- Mul
+- MulAssign
+- Neg
+- Not
+- Shl
+- ShlAssign
+- Shr
+- ShrAssign
+- Sub
+- SubAssign
+- Sub
+
+
+For secret_bool, implement the following methods:
+- n/a
+And the following traits
+- BitAnd
+- BitAndAssign
+- BitOr
+- BitOrAssign
+- BitXor
+- BitXorAssign
+- Clone
+- Copy
+- Default
+- Not
+
+We will also need to define a trait Classify<T> for T and method declassify. Classify will take a non secret integer or boolean and return a secret integer or boolean. `declassify` will consume a secret integer returning a non-secret integer.
+
+We will need to define new comparison traits that run in constant time and return a secret_bool:
+- SecretEq
+- SecretOrd
+For reference, see the subtle crate [2] .
+
+[2] https://docs.rs/subtle/2.2.2/subtle/
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Because secret integers prohibit the use of certain operations and compiler optimizations, there will be a performance impact on code that uses them. However, this is intentional.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main alternative to this design is to handle this via crates such as secret_integers [3]. However, without compiler integration, the compiler could optimize out the guarantees that have been claimed at the source level. For example, the compiler could use non constant time instructions such as divide. Using these primitives correctly from a high-level language will typically require careful management of memory allocation, flagging relevant memory as containing secret for the operating system, hardware, or other system level components.
+Other alternatives would require developers to use handwritten assembly to ensure that only constant time operations are used.
+
+[3] https://docs.rs/secret_integers/0.1.5/secret_integers/
+ 
+
+# Prior art
+[prior-art]: #prior-art
+- https://docs.rs/secret_integers/0.1.5/secret_integers/
+- https://docs.rs/subtle/2.2.2/subtle/
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+Out of scope: Memory zeroing (beyond implementing the drop trait) and register spilling
+
+There is an in progress RFC for LLVM that will be required for the secret type guarantees to be fully guaranteed throughout the compilation process.
```

Comment:
What about Cranelift?

Reply:
If Cranelift were to add support for this, the earlier the better.

It's going to be a monumental effort for LLVM, not because the work is complicated - it's simple and mundane, but there is going to be a lot of it because it touches every layer of code generation.

Reply:
Great. This should be elaborated upon in the text.

Reply:
cc https://github.com/bytecodealliance/cranelift/issues/1327

Reply:
@bjorn3 aah nice!

Reply:
I think for this a special type may be better though to also prevent any branches, non constant time multiplication and other ways of leaking it.